### PR TITLE
feat(gatsby-plugin-sitemap): create sitemap index

### PR DESCRIPTION
**Pull request by tomoyukikashiro**
## Description

Now `gatsby-plugin-sitemap` can not create [sitemap index](https://github.com/ekalinin/sitemap.js#example-of-sitemap-index).
This PR allow us to create it by passing additional option to that plugin.

Also this plugin did not have any tests so I added them.

## Related Issues

#12100 

## Details

```javascript
// In your gatsby-config.js
siteMetadata: {
  siteUrl: `https://www.example.com`,
},
plugins: [
  {
    resolve: `gatsby-plugin-sitemap`,
    options: {
      sitemapSize: 5000
    }
  }
]
```

Above is the minimal configuration to split large sitemap.
When number of URL in sitemap is more than 5000 plugin will create sitemap (e.g. `sitemap-0.xml`, `sitemap-1.xml`) and index (e.g. `sitemap.xml`) files.

### CODE REVIEW

**Comment by tomoyukikashiro** [comment]
@pieh I assigned you as reviewer 🙇 because github suggested that you reviewed these file recently... I did not know what to do after making PR ( assigned someone as reviewer by myself or just wait).

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 1]
```diff
@@ -26,7 +40,37 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
...
+        `${rest.sitemapName || `sitemap`}-index.xml`
+      )
+      const sitemapIndexOptions = {
+        ...rest,
+        ...{
+          hostname: hostname || withoutTrailingSlash(siteUrl),
+          targetFolder: distDir,
+          urls,
+          callback: error => {
+            if (error) reject()
```
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js]: maybe throw or reject with the error?
```suggestion
            if (error) throw new Error(error)
```

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 2]
```diff
@@ -26,7 +40,37 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
...
+    } = queryRecords
+    return new Promise((resolve, reject) => {
+      // sitemapv-index.xml is default file name. (https://git.io/fhNgG)
+      const indexFilePath = path.join(
+        distDir,
+        `${rest.sitemapName || `sitemap`}-index.xml`
+      )
+      const sitemapIndexOptions = {
+        ...rest,
+        ...{
```
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js]:
```suggestion
```
you can remove this line

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 3]
```diff
@@ -26,7 +40,37 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
...
+          hostname: hostname || withoutTrailingSlash(siteUrl),
+          targetFolder: distDir,
+          urls,
+          callback: error => {
+            if (error) reject()
+            renameFile(indexFilePath, saved).then(resolve)
+          },
+        },
```
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js]:
```suggestion
        },
```
same here as we remove line `...{`

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 4]
```diff
@@ -26,7 +40,37 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
     excludeOptions,
     pathPrefix
   )
-  serialize(queryRecords).forEach(u => map.add(u))
+  const urls = serialize(queryRecords)
 
-  return await writeFile(saved, map.toString())
+  if (!(rest.sitemapSize && urls.length > rest.sitemapSize)) {
```
Comment (wardpeet) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js]: I have a hard time wrapping my head around `!(conditions)` probably just me.

```suggestion
  if (!rest.sitemapSize || urls.length <= rest.sitemapSize) {
```

**Comment by wardpeet** [comment]
I also tested the plugin and works!

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 5]
```diff
@@ -9,13 +15,21 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
...
+    targetFolder,
+    ...rest
+  } = {
     ...defaultOptions,
     ...options,
   }
 
-  const map = sitemap.createSitemap(rest)
-  const saved = path.join(publicPath, output)
+  const distDir = targetFolder || publicPath
```
[the remark this answers is not in the document]
Comment (tomoyukikashiro) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: Yes... it's only for test. should I introduce `options.publichPath (default is /public )` or mock [fs.createWriteStream](https://github.com/ekalinin/sitemap.js/blob/master/lib/sitemap.js#L376-L413) in sitemap module ?

**Comment by tomoyukikashiro** [comment]
@wardpeet Thanks for your review 👍 I have one question !
↓
https://github.com/gatsbyjs/gatsby/pull/12239#discussion_r262473766

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (wardpeet) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: if you could, I would say mock it

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 6]
```diff
@@ -26,7 +40,35 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
     excludeOptions,
     pathPrefix
   )
-  serialize(queryRecords).forEach(u => map.add(u))
+  const urls = serialize(queryRecords)
+
+  if (!rest.sitemapSize || urls.length <= rest.sitemapSize) {
+    const map = sitemap.createSitemap(rest)
+    serialize(queryRecords).forEach(u => map.add(u))
```
Comment (muescha) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js]: why not `urls`?
```suggestion
    urls.forEach(u => map.add(u))
```

Comment (tomoyukikashiro) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: @muescha Thanks ! I've just fixed it. 👍

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomoyukikashiro) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: @wardpeet I've just fixed it to use `mock`. 👍

**Comment by wardpeet** [comment]
Works like a charm, thank you!

**Comment by sidharthachatterjee** [comment]
Published in gatsby-plugin-sitemap@2.0.8

File: packages/gatsby-plugin-sitemap/src/gatsby-node.js [hunk 7]
```diff
@@ -26,7 +31,35 @@ exports.onPostBuild = async ({ graphql, pathPrefix }, pluginOptions) => {
...
+    return writeFile(saved, map.toString())
+  }
+
+  const {
+    site: {
+      siteMetadata: { siteUrl },
+    },
+  } = queryRecords
+  return new Promise(resolve => {
+    // sitemapv-index.xml is default file name. (https://git.io/fhNgG)
```
Comment (muescha) [inline packages/gatsby-plugin-sitemap/src/gatsby-node.js]:
```suggestion
    // sitemap-index.xml is default file name. (https://git.io/fhNgG)
```

`sitemapv-index` -> `sitemap-index`?

Comment (wardpeet) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: @muescha it's already merged, could you add a new pr? :)

Comment (muescha) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: ok done -> #12512

Comment (tomoyukikashiro) [reply packages/gatsby-plugin-sitemap/src/gatsby-node.js]: @muescha @wardpeet Thanks for that 😓

**Pull request merged**